bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandler.java

```diff
@@ -0,0 +1,167 @@
+    public void handleCommand(ChannelUID channelUID, Command command) {
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(ProwlConfiguration.class);
+        updateStatus(ThingStatus.UNKNOWN);
+
+        statusFuture = scheduler.scheduleWithFixedDelay(() -> {
+            updateStatus();
+        }, 0, config.refresh, TimeUnit.MINUTES);
+    }
+
+    private void updateStatus() {
+        if (keyVerificationSucceeded(config.apiKey)) {
+            updateStatus(ThingStatus.ONLINE);
+        } else {
+            updateStatus(ThingStatus.OFFLINE);
+        }
+    }
+
+    @Override
+    public void dispose() {
+        ScheduledFuture<?> localPollFuture = statusFuture;
+        if (localPollFuture != null && !localPollFuture.isCancelled()) {
+            localPollFuture.cancel(true);
+        }
+        super.dispose();
+    }
+
+    private boolean keyVerificationSucceeded(String apiKey) {
+        try {
+            ContentResponse response = httpClient.GET(PROWL_VERIFY_URI + "?apikey=" + apiKey);
+            String resp = response.getContentAsString();
+            if (logger.isTraceEnabled()) {
+                logger.trace("verify response: {}", resp);
+            }
+            if (resp.contains("<success code=\"200\"")) {
+                updateFreeMessages(resp);
+                return true;
+            } else {
+                return false;
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        } catch (ExecutionException e) {
+            logger.debug("error during calling uri: {}", PROWL_ADD_URI, e);
+        } catch (TimeoutException e) {
+            logger.debug("timeout during calling uri: {}", PROWL_ADD_URI, e);
+        }
+        return false;
+    }
+
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Collections.singletonList(ProwlActions.class);
+    }
+
+    public void pushNotification(@Nullable String event, @Nullable String description) {
+        if (event == null || description == null) {
+            logger.debug("Cannot push message with null event or null description");
+            return;
+        }
+
+        logger.debug("Pushing an event: {} with desc: {}", event, description);
+        try {
+            ContentResponse response = httpClient.POST(PROWL_ADD_URI).timeout(5, TimeUnit.SECONDS)
+                    .content(
+                            new StringContentProvider("apikey=" + config.apiKey + "&application=" + config.application
+                                    + "&event=" + event + "&description=" + description),
+                            "application/x-www-form-urlencoded; charset=UTF-8")
+                    .send();
+            String resp = response.getContentAsString();
+            updateFreeMessages(resp);
+            if (logger.isTraceEnabled()) {
+                logger.trace("add response: {}", resp);
+            }
```

Comment:
The logger only does string substitution if the log level matches. So it should have not much impact on performance.

```suggestion
            logger.trace("add response: {}", resp);
```

Reply:
fixed